sympy/functions/elementary/piecewise.py

```diff
@@ -1248,4 +1271,205 @@ def piecewise_simplify(expr, **kwargs):
+            p = -len(args)  # Store previous index, first time there is no previous index.
+            for v in pos:
+                if v:  # If zero => first item, no previous for sure
+                    newstep = v - p
+                    if minstep > newstep:  # Smaller than earlier best?
+                        minstep = newstep
+                        c0 = p
+                        c1 = v
+                p = v
+        # Compute cummulative condition for bypassing intermediate expressions
+        cumcond = And(*[args[k][1] for k in range(c0 + 1, c1)])
+        lasttrue = (args[-1][1] == S.true)  # Remember if last condition is True
+        # Update condition
+        args[c0] = (args[c0][0],
+                    (args[c0][1] | (~cumcond & args[c1][1])).simplify())
+        del args[c1]
+        # Set final condition to true
+        if lasttrue:
+            args[-1] = (args[-1][0], S.true)
+        # Go again
+        return _piecewise_simplify_identical_expressions(args)
+    return args
+
+def _piecewise_replace_equal_expressions(args):
+    d = defaultdict(list)
+    newargs = []
+    uniqueexpr = []
+    for expr, cond in args:
+        unique = True
+        for uexpr in uniqueexpr:
+            if uexpr.equals(expr):
+                unique = False
+                newargs.append((expr, cond))
+                break
+        if unique:
+            newargs.append((expr, cond))
+    return newargs
+
+
+def _piecewise_simplify_conditions(args, measure, ratio):
+    """
+    Propagate earlier conditions to latter to see if they simplify.
+    """
+    def shorter(newcond, oldcond):
+        return newcond if measure(newcond) < ratio*measure(oldcond) else oldcond
+
+    cummulatedcond = S.false
+    newargs = []
+    prevcond = S.false
+    prevnewcond = S.false
+    for expr, cond in args:
+        newcond = (cond & ~cummulatedcond).simplify()
+        shortcond = shorter(newcond, cond)
+        if shortcond != False and shortcond not in [prevcond, prevnewcond]:
+            newargs.append((expr, shortcond))
+        cummulatedcond = (cummulatedcond | shortcond).simplify()
+        prevcond, prevnewcond = cond, newcond
+    return newargs
+
+
+def piecewise_simplify(expr, **kwargs):
+    """
+    Simplify a Piecewise expression
+    """
+
+    if not isinstance(expr, Piecewise):
+        return expr
+
+    expr = piecewise_simplify_arguments(expr, **kwargs)
+    args = list(expr.args)
+    measure = kwargs['measure']
+    ratio = kwargs['ratio']
+
+    args = _piecewise_simplify_and_eq(args)
+    args = _piecewise_simplify_eq_next_segment(args)
+    args = _piecewise_simplify_or_and_eq(args, measure, ratio)
+    args = _piecewise_simplify_ne(args)
+    args = _piecewise_collapse_arguments(args)
+    args = _piecewise_simplify_conditions(args, measure, ratio)
+    args = _piecewise_replace_equal_expressions(args)
+    args = _piecewise_simplify_identical_expressions(args)
+
     return Piecewise(*args)
```

Comment:
Could this use `evaluate=False` or do we need to reevaluate after simplifying?

Reply:
I'll have to think a bit about it. It is not clear which order one should do the simplification steps in and how many times. Most of the evaluation magic is taken care of in the `collapse` magic anyway.